File: clients/python/transcript_create_client/rate_limiter.py
"""Client-side rate limiting."""

import asyncio
import time
from typing import Optional


class RateLimiter:
    """Token bucket rate limiter for client-side rate limiting."""

    def __init__(
        self,
        requests_per_second: float = 10.0,
        burst_size: Optional[int] = None,
    ) -> None:
        """Initialize rate limiter.

        Args:
            requests_per_second: Maximum requests per second
            burst_size: Maximum burst size (defaults to requests_per_second)
        """
        self.rate = requests_per_second
        self.burst_size = burst_size or int(requests_per_second)
        self.tokens = float(self.burst_size)
        self.last_update = time.monotonic()
        self._lock = asyncio.Lock()

    async def acquire(self) -> None:
        """Acquire a token, waiting if necessary."""
        async with self._lock:
            while True:
                now = time.monotonic()
                elapsed = now - self.last_update
                self.last_update = now

                # Add tokens based on elapsed time
                self.tokens = min(self.burst_size, self.tokens + elapsed * self.rate)

                if self.tokens >= 1.0:
                    self.tokens -= 1.0
                    return

                # Calculate wait time for next token
                wait_time = (1.0 - self.tokens) / self.rate
                await asyncio.sleep(wait_time)

    def reset(self) -> None:
        """Reset rate limiter to full capacity."""
        self.tokens = float(self.burst_size)
        self.last_update = time.monotonic()


class AdaptiveRateLimiter(RateLimiter):
    """Rate limiter that adapts based on 429 responses."""

    def __init__(
        self,
        initial_requests_per_second: float = 10.0,
        min_requests_per_second: float = 1.0,
        max_requests_per_second: float = 100.0,
        increase_factor: float = 1.1,
        decrease_factor: float = 0.5,
    ) -> None:
        """Initialize adaptive rate limiter.

        Args:
            initial_requests_per_second: Starting rate
            min_requests_per_second: Minimum allowed rate
            max_requests_per_second: Maximum allowed rate
            increase_factor: Factor to increase rate on success
            decrease_factor: Factor to decrease rate on 429
        """
        super().__init__(initial_requests_per_second)
        self.min_rate = min_requests_per_second
        self.max_rate = max_requests_per_second
        self.increase_factor = increase_factor
        self.decrease_factor = decrease_factor
        self.success_count = 0
        self.success_threshold = 10  # Increase rate after N successes

    async def on_success(self) -> None:
        """Called after successful request."""
        async with self._lock:
            self.success_count += 1

            if self.success_count >= self.success_threshold:
                # Gradually increase rate
                new_rate = min(self.rate * self.increase_factor, self.max_rate)
                if new_rate != self.rate:
                    self.rate = new_rate
                    self.success_count = 0

    async def on_rate_limit(self) -> None:
        """Called when rate limit is hit."""
        async with self._lock:
            # Decrease rate immediately
            self.rate = max(self.rate * self.decrease_factor, self.min_rate)
            self.success_count = 0

File: clients/javascript/src/rate-limiter.ts
/**
 * Client-side rate limiting
 */

/**
 * Token bucket rate limiter
 */
export class RateLimiter {
  protected rate: number;
  private burstSize: number;
  private tokens: number;
  private lastUpdate: number;

  constructor(requestsPerSecond: number, burstSize?: number) {
    this.rate = requestsPerSecond;
    this.burstSize = burstSize || Math.floor(requestsPerSecond);
    this.tokens = this.burstSize;
    this.lastUpdate = Date.now();
  }

  /**
   * Acquire a token, waiting if necessary
   */
  async acquire(): Promise<void> {
    while (true) {
      const now = Date.now();
      const elapsed = (now - this.lastUpdate) / 1000;
      this.lastUpdate = now;

      // Add tokens based on elapsed time
      this.tokens = Math.min(this.burstSize, this.tokens + elapsed * this.rate);

      if (this.tokens >= 1) {
        this.tokens -= 1;
        return;
      }

      // Calculate wait time for next token
      const waitTime = ((1 - this.tokens) / this.rate) * 1000;
      await new Promise((resolve) => setTimeout(resolve, waitTime));
    }
  }

  /**
   * Reset rate limiter to full capacity
   */
  reset(): void {
    this.tokens = this.burstSize;
    this.lastUpdate = Date.now();
  }
}

/**
 * Adaptive rate limiter that adjusts based on 429 responses
 */
export class AdaptiveRateLimiter extends RateLimiter {
  private currentRate: number;
  private minRate: number;
  private maxRate: number;
  private increaseFactor: number;
  private decreaseFactor: number;
  private successCount: number;
  private successThreshold: number;

  constructor(
    initialRequestsPerSecond: number,
    options: {
      minRate?: number;
      maxRate?: number;
      increaseFactor?: number;
      decreaseFactor?: number;
    } = {}
  ) {
    super(initialRequestsPerSecond);
    this.currentRate = initialRequestsPerSecond;
    this.minRate = options.minRate || 1;
    this.maxRate = options.maxRate || 100;
    this.increaseFactor = options.increaseFactor || 1.1;
    this.decreaseFactor = options.decreaseFactor || 0.5;
    this.successCount = 0;
    this.successThreshold = 10;
  }

  /**
   * Called after successful request
   */
  onSuccess(): void {
    this.successCount++;

    if (this.successCount >= this.successThreshold) {
      // Gradually increase rate
      const newRate = Math.min(this.currentRate * this.increaseFactor, this.maxRate);
      if (newRate !== this.currentRate) {
        this.currentRate = newRate;
        this.rate = newRate; // Update parent class rate
        this.successCount = 0;
      }
    }
  }

  /**
   * Called when rate limit is hit
   */
  onRateLimit(): void {
    // Decrease rate immediately
    this.currentRate = Math.max(this.currentRate * this.decreaseFactor, this.minRate);
    this.rate = this.currentRate; // Update parent class rate
    this.successCount = 0;
  }
}
